Self as generic trait parameter is not properly ignored by Self: Sized bound

### This Compiles
```Rust
struct Foo;
trait Bar{
    fn some_method(&self) -> impl SomeTrait where Self: Sized;
}
trait SomeTrait{}
fn main() {
    let obj : &dyn Bar;
}
```
### This Also Compiles
```Rust
struct Foo;
trait Bar{
    fn some_method(&self) -> &dyn SomeTrait<Self> where Self: Sized;
}
trait SomeTrait<T>{}
fn main() {
    let obj : &dyn Bar;
}
```
### But This..

```Rust
struct Foo;
trait Bar{
    fn some_method(&self) -> impl SomeTrait<Self> where Self: Sized;
}
trait SomeTrait<T>{}
fn main() {
    let obj : &dyn Bar;
}
```
### Show This Error output
```
error[E0038]: the trait `Bar` is not dyn compatible
 --> src\main.rs:7:15
  |
7 |     let obj : &dyn Bar;
  |               ^^^^^^^^ `Bar` is not dyn compatible
  |
note: for a trait to be dyn compatible it needs to allow building a vtable
      for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
 --> src\main.rs:3:35
  |
2 | trait Bar{
  |       --- this trait is not dyn compatible...
3 |     fn some_method(&self) -> impl SomeTrait<Self> where Self: Sized;
  |                                   ^^^^^^^^^^^^^^^ ...because it uses `Self` as a type parameter
```
Even though Self : Sized. This should be valid in rust.

version:
rustc 1.86.0-nightly (a567209da 2025-02-13)

--- a/compiler/rustc_trait_selection/src/traits/dyn_compatibility.rs
+++ b/compiler/rustc_trait_selection/src/traits/dyn_compatibility.rs
@@ -187,7 +187,10 @@ fn predicates_reference_self(
 fn bounds_reference_self(tcx: TyCtxt<'_>, trait_def_id: DefId) -> SmallVec<[Span; 1]> {
     tcx.associated_items(trait_def_id)
         .in_definition_order()
+        // We're only looking at associated type bounds
         .filter(|item| item.kind == ty::AssocKind::Type)
+        // Ignore GATs with `Self: Sized`
+        .filter(|item| !tcx.generics_require_sized_self(item.def_id))
         .flat_map(|item| tcx.explicit_item_bounds(item.def_id).iter_identity_copied())
         .filter_map(|(clause, sp)| {
             // Item bounds *can* have self projections, since they never get

--- a/tests/ui/dyn-compatibility/associated_type_bound_mentions_self.rs
+++ b/tests/ui/dyn-compatibility/associated_type_bound_mentions_self.rs
@@ -1,0 +1,14 @@
+// Ensure that we properly ignore the `B<Self>` associated type bound on `A::T`
+// since that associated type requires `Self: Sized`.
+
+//@ check-pass
+
+struct X(&'static dyn A);
+
+trait A {
+    type T: B<Self> where Self: Sized;
+}
+
+trait B<T> {}
+
+fn main() {}

